### PR TITLE
Enable Azure Data Studio's database selector to behave like SSMS'

### LINKED ISSUE #16659: Database name's text box's text isn't used whenever a query is run
Issue Type: <b>Bug</b>

1. Open Azure Data Studio and connect to a server with multiple databases.
2. Select a database from the dropdown in header. Actually, even after typing it, you will have to click the database's name in the dropdown.
image

3. Now(preferably) run a select command on a table in the database we just selected. For this illustration I selected 'master' database and have a table 'TblBugTest' in it.  It is supposed to run perfectly as expected.
BE0DAB87-283A-4EC7-90D2-935671E78050

4. Now I have another database 'APIRND', type your second database's name in the database name text box accurately but do not press enter or click the name in the dropdown.
5. Now for my illustration, 'APIRND' doesn't have 'TblBugTest' table, but when we will press run, the command will execute successfully, moreover it will run that query on the last dropdown clicked database(i.e. master in this case).
B6004FCF-A760-4F19-920B-8EEC097476C0

6. Now, type 'APIRND' in the database name text box but this time click on this database's name in the dropdown.
7. Now, since the table is not in the second database , it will not execute successfully.
image

My assumption is that clicking on the database name on the dropdown updates a variable and that variable is used as the target database.  In my opinion, either the database name's text box should always reflect that variable's value or everytime the query is run or the target database's name should be fetched from the database name's texbox.

I know the active database’s name is in the bottom right corner but it can still cause confusion to those who are migrating to Azure Data Studio on UNIX or windows systems and may also drive them to reach wrong conclusions while handling their data.

Azure Data Studio version: azuredatastudio 1.31.1 (65fb22cc7c36db9c53af1ed2fdbdf48f66c682be, 2021-07-28T21:03:21.018Z)
OS version: Windows_NT x64 10.0.19043

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-8300H CPU @ 2.30GHz (8 x 2304)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|7.86GB (2.20GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (5)</summary>

Extension|Author (truncated)|Version
---|---|---
admin-pack|Mic|0.0.2
agent|Mic|0.48.0
dacpac|Mic|1.8.0
import|Mic|1.3.1
profiler|Mic|0.11.0

</details>

### PR DESCRIPTION
**Pull request by lewis-sanchez**
This PR fixes #16659 

**What is the PR?**
This PR will fix an issue around the database selector not updating to the selected database when a database name is typed or pasted into the editable drop down when focus is taken away from the drop down.

**How was fix determined?**
The fix was determined by observing the behavior of the database drop down when the enter key is pressed on the keyboard and seeing that an "onValueChange" event is fired off to update the selected database with the input's value. By registering to the dropdown's "onBlur event listener, a manually entered database name will now be committed when focus is moved away from the DB selector dropdown. If a database name is entered incorrectly, then the DB selector will have a red outline and will provide an explanation for the error.

**Verification**
 - Open Azure Data Studio and connect to a server with multiple databases.
 - Expand the database drop down list, and manually type in any of the available database names.
 - Without pressing enter or clicking on the database name in the drop down list, click anywhere outside the drop down and drop down list to take it out of focus.
 
Expected: The ADS window title bar will be updated with the manually entered database name. (msdb, in following image)
image

### CODE REVIEW

File: src/sql/base/parts/editableDropdown/browser/dropdown.ts
Comment (Charles-Gagnon) [inline src/sql/base/parts/editableDropdown/browser/dropdown.ts:155]: We already have [fireOnTextChange](https://github.com/microsoft/azuredatastudio/blob/main/src/sql/base/parts/editableDropdown/browser/dropdown.ts#L77) that controls whether the onValueChange event is fired whenever the text changes. Did you look into using that?

In general we should be careful about making changes like this to core components. Dropdowns are used all across the app and changing the behavior like this could result in issues in other parts of the app. A change like this doesn't necessarily seem like something all consumers would want to have be the default behavior.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: src/sql/base/parts/editableDropdown/browser/dropdown.ts
Comment (Charles-Gagnon) [inline src/sql/base/parts/editableDropdown/browser/dropdown.ts:150]: This comment specifically mentions the database selector but this class is for a generic dropdown - not only the database selector.

File: src/sql/workbench/contrib/query/browser/queryActions.ts
Comment (lewis-sanchez) [reply src/sql/workbench/contrib/query/browser/queryActions.ts:613]: When the user enters an invalid db name, they don't see any error messages until the drop down loses focus, but they do see the list of valid db names.
image

Once the db selector is no longer in focus and contains the invalid db name, the user will see a red outline around the selector:
image

The error message appears when the user highlights the incorrect db name in the selector:
image

**Review by Charles-Gagnon — APPROVED** [review]

**Comment by Charles-Gagnon** [comment]
@lewis-sanchez Looks like there's a merge conflict now. Could you please merge in the latest changes from main and resolve the conflict so that other people can review this?

**Comment by lewis-sanchez** [comment]
@Charles-Gagnon, finished resolving merge conflicts.

File: src/sql/base/parts/editableDropdown/browser/dropdown.ts
Comment (alanrenmsft) [reply src/sql/base/parts/editableDropdown/browser/dropdown.ts]: @lewis-sanchez have you tried to set the fireOnTextChange option to true?

Comment (alanrenmsft) [reply src/sql/base/parts/editableDropdown/browser/dropdown.ts]: are you not using that because it will be fired multiple times?

Comment (Charles-Gagnon) [reply src/sql/base/parts/editableDropdown/browser/dropdown.ts]: Yes, he tried it. And I don't think that's the correct answer because yes, it would be fired every single time the text changed. We only really care about getting the value when the user either selects a new value or tabs out of the select box so onBlur seems like the better option to use.

File: src/sql/workbench/contrib/query/browser/queryActions.ts
Comment (alanrenmsft) [inline src/sql/workbench/contrib/query/browser/queryActions.ts:600]: call to the this.databaseSelected method will trigger a changeDatabase operation which is expensive, we should modify it to only do it when the database is actually changed, especially when the user is only navigating using keyboard.

File: src/sql/base/parts/editableDropdown/browser/dropdown.ts
Comment (alanrenmsft) [reply src/sql/base/parts/editableDropdown/browser/dropdown.ts]: agreed, using onBlur sounds good to me.

Comment (lewis-sanchez) [reply src/sql/base/parts/editableDropdown/browser/dropdown.ts]: @alanrenmsft when I set the fireOnTextChange option to true, that broke the filtering aspect of the db selector drop down because it would show all of them. With the onBlur approach that isn't the case and filtering works as expected.

File: src/sql/workbench/contrib/query/browser/queryActions.ts
Comment (lewis-sanchez) [reply src/sql/workbench/contrib/query/browser/queryActions.ts:600]: I will go back and make the proper updates to this.

File: src/sql/workbench/contrib/query/browser/queryActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/browser/queryActions.ts:601]: We should be doing this check inside databaseSelected instead so we avoid switching the DB unless the db actually changed in all scenarios.

File: src/sql/workbench/contrib/query/browser/queryActions.ts
Comment (lewis-sanchez) [reply src/sql/workbench/contrib/query/browser/queryActions.ts:600]: I moved the requested check to the databaseSelected method, so that we're only committing the change when the DB name has changed.

File: src/sql/workbench/contrib/query/browser/queryActions.ts
Comment (alanrenmsft) [inline src/sql/workbench/contrib/query/browser/queryActions.ts:670]: we should use dbName here instead of this._dropdown.value

**Review by alanrenmsft — APPROVED** [review]

File: src/sql/workbench/contrib/query/browser/queryActions.ts
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/browser/queryActions.ts:670]: Would suggest putting this at the top, right below the !dbName check so that we avoid having to call getConnectionProfile if we can.

Comment (lewis-sanchez) [reply src/sql/workbench/contrib/query/browser/queryActions.ts]: @Charles-Gagnon, moved the check below the !dbName check to avoid calling getConnectionProfile.

**Review by Charles-Gagnon — APPROVED** [review]

**Comment by Charles-Gagnon** [comment]
@lewis-sanchez Are you waiting for something before merging this in?

**Comment by lewis-sanchez** [comment]
@Charles-Gagnon, I was waiting for @kburtram to review because I thought I needed all reviewers to review the PR to merge. If not, I can merge it in now.

**Review by kburtram — APPROVED** [review]

**Pull request merged**
